### PR TITLE
Add support for AAD auth in data plane ops

### PR DESCRIPTION
**Pull request by PratibhaShrivastav18**
# Description

Adding support for AAD (aka Microsoft entra token auth) for data plane operations , viz. `get-credentials` and `invoke`
[Here ](https://github.com/MicrosoftDocs/azure-docs-pr/pull/271892)is the doc update PR for the same.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_endpoint/online_endpoint.py [hunk 1]
```diff
@@ -625,3 +626,22 @@ def _to_rest_object(self) -> RestEndpointAuthToken:
...
+    """Endpoint aad token.
+
+    :ivar access_token: Access token for aad authentication.
+    :vartype access_token: str
+    :ivar expiry_time_utc: Access token expiry time (UTC).
+    :vartype expiry_time_utc: float
+    """
+
+    def __init__(self, obj: AccessToken):
+        """Constuctor for Endpoint aad token.
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_endpoint/online_endpoint.py:641]: Small typo: *Constructor

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_endpoint/online_endpoint.py [hunk 2]
```diff
@@ -625,3 +626,22 @@ def _to_rest_object(self) -> RestEndpointAuthToken:
             refresh_after_time_utc=self.refresh_after_time_utc,
             token_type=self.token_type,
         )
+
+
+class EndpointAadToken:
+    """Endpoint aad token.
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_endpoint/online_endpoint.py:632]: Thanks for adding docstrings!!

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_online_endpoint_operations.py [hunk 3]
```diff
@@ -384,6 +391,9 @@ def _get_online_credentials(
                 **self._init_kwargs,
             )
 
+        if auth_mode is not None and auth_mode.lower() == AAD_TOKEN:
```
Comment (diondrapeck) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_online_endpoint_operations.py:394]: I like this design of encasing it in a class.

**Review by diondrapeck — COMMENTED** [review]
This looks really great!

- Could you also update the changelog file to include this change for the next release (1.16.0)? azure-sdk-for-python/sdk/ml/azure-ai-ml/CHANGELOG.md
- The only failures you're seeing are in the Analyze step of the PR gates. Analyze contains linting and style checks. Right now, mypy (a python type checker) is failing. You can fix both errors by 1. Changing the return type of `get_keys` to include the new class you're returning from `_get_online_credentials` and 2. Adding a check for if `self._credentials` is None. We could add an exception message here if it's None to say that they need to make sure they're signed in with their MLClient. e.g.

```python
if auth_mode is not None and auth_mode.lower() == AAD_TOKEN:
    if self._credentials:
        return EndpointAadToken(self._credentials.get_token(*_resource_to_scopes(AAD_TOKEN_RESOURCE_ENDPOINT)))
    raise Exception("No credential available. Initialize credential using the `MLClient` if using SDK or `az login` if using the CLI")
```

**Review by diondrapeck — APPROVED** [review]

**Review by TajiHarrisMicrosoft — APPROVED** [review]

File: sdk/ml/azure-ai-ml/CHANGELOG.md [hunk 4]
```diff
@@ -1,14 +1,9 @@
 # Release History
 
-## 1.16.0 (unreleased)
+## 1.16.0 (2024-04-22)
```
Comment (luigiw) [inline sdk/ml/azure-ai-ml/CHANGELOG.md:3]: Can you revert this? This date is supposed to be when the new SDK version is released.

**Comment by luigiw** [comment]
Please address above comment before merge.

File: sdk/ml/azure-ai-ml/CHANGELOG.md (hunk 4 above)
Comment (PratibhaShrivastav18) [reply sdk/ml/azure-ai-ml/CHANGELOG.md]: Reverted. Thanks for letting me know.

**Pull request merged**